src/libsyntax/ext/expand.rs

```diff
@@ -707,6 +778,28 @@ pub fn expand_crate(mut cx: ExtCtxt,
     return (ret, cx.syntax_env.names);
 }
 
+// Expands crate using supplied MacroExpander - allows for
+// non-standard expansion behaviour (e.g. step-wise).
+pub fn expand_crate_with_expander(expander: &mut MacroExpander,
+                                  user_exts: Vec<NamedSyntaxExtension>,
+                                  mut c: Crate) -> (Crate, HashSet<Name>) {
+    init_cx(expander.cx, user_exts, &c);
+
+    let items = SmallVector::many(c.module.items);
+    expander.load_macros(&items);
+    c.module.items = items.into();
+
+    let err_count = expander.cx.parse_sess.span_diagnostic.err_count();
+    let mut ret = expander.fold_crate(c);
+    ret.exported_macros = expander.cx.exported_macros.clone();
+
+    if expander.cx.parse_sess.span_diagnostic.err_count() > err_count {
+        expander.cx.parse_sess.span_diagnostic.abort_if_errors();
+    }
+
+    return (ret, expander.cx.syntax_env.names.clone());
```

Comment:
To avoid this clone, I would just return `ret` here.
If a tool wants to access to `names` (seems unlikely), it can use `cx.syntax_env.names` directly.